### PR TITLE
React Pages Migration: New/Edit Audio Response Question

### PR DESCRIPTION
**Pull request by bivanalhar**
Part of #5098.

| Before | After |
| - | - |
| before | after |

Some other minor differences:
- move the commonly used fields among all types of questions (Question Details, Skills, and Maximum Grade) to Common directory, so that next time the question form is ported or refactored, this page can be called multiple times for multiple purposes

### CODE REVIEW

File: app/views/course/assessment/question/voice_responses/new.json.jbuilder [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+json.partial! 'form', locals: {
+  question: @voice_response_question,
```
Comment (ekowidianto) [inline app/views/course/assessment/question/voice_responses/new.json.jbuilder:3]: Do we need to pass the variables from line 3 to 5?

File: app/views/course/assessment/question/voice_responses/edit.json.jbuilder [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+question = @voice_response_question
+question_assessment = @question_assessment
+
+json.partial! 'form', locals: {
+  question: question,
```
Comment (ekowidianto) [inline app/views/course/assessment/question/voice_responses/edit.json.jbuilder:6]: Do we need line 6 to 8?

File: app/views/course/assessment/question/voice_responses/edit.json.jbuilder [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  question_assessment: question_assessment,
+  new_question: false,
+  course: current_course
+}
+
+json.question do
+  json.partial! 'course/assessment/question/form', question: question
+  json.skillIds question_assessment.skill_ids
```
Comment (ekowidianto) [inline app/views/course/assessment/question/voice_responses/edit.json.jbuilder:14]: You can move `skillIds` to `'course/assessment/question/form'` too. Dont forget to remove the same line in `app/views/course/assessment/question/multiple_responses/edit.json.jbuilder` after `json.skillIds` has been abstracted

File: client/app/api/course/Assessment/Question/index.ts [hunk 4]
```diff
@@ -1,9 +1,11 @@
 import McqMrqAPI from './McqMrq';
 import ScribingQuestionAPI from './Scribing';
+import VoiceAPI from './Voice';
 
 const QuestionAPI = {
   mcqMrq: new McqMrqAPI(),
   scribing: new ScribingQuestionAPI(),
+  voice: new VoiceAPI(),
```
Comment (ekowidianto) [inline client/app/api/course/Assessment/Question/index.ts:8]: nit: rename to VoiceResponsesAPI(). Similarly, rename `Voice.ts` to `VoiceResponses.ts` and everything in that file.

File: client/app/api/course/Assessment/Question/Voice.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+export default class VoiceAPI extends BaseAPI {
+  get #urlPrefix(): string {
+    return `/courses/${this.courseId}/assessments/${this.assessmentId}/question/voice_responses`;
+  }
+
+  fetchNewVoice(): APIResponse<VoiceFormData<'new'>> {
+    return this.client.get(`${this.#urlPrefix}/new`);
+  }
+
+  fetchEdit(id: number): APIResponse<VoiceFormData<'edit'>> {
```
Comment (ekowidianto) [inline client/app/api/course/Assessment/Question/Voice.ts:19]: nit: `fetchEditVoice`

File: client/app/types/course/assessment/question/voice-responses.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { AvailableSkills, QuestionFormData } from '../questions';
+
+type OptionalIfNew<T extends 'new' | 'edit'> = T extends 'new'
```
Comment (ekowidianto) [inline client/app/types/course/assessment/question/voice-responses.ts:3]: This seems to be repeated in `multiple-responses.ts`. Can you abstract this to `questions.ts`?

File: client/app/types/course/assessment/question/voice-responses.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  : never;
+
+export interface VoiceData<T extends 'new' | 'edit' = 'edit'> {
+  question: QuestionFormData | OptionalIfNew<T>;
+}
+
+export interface VoiceFormData<T extends 'new' | 'edit' = 'edit'>
+  extends VoiceData<T>,
+    AvailableSkills {
+  allowRandomization: boolean;
```
Comment (ekowidianto) [inline client/app/types/course/assessment/question/voice-responses.ts:14]: What is this?

File: client/app/bundles/course/assessment/question/voice-responses/operations.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import {
+  VoiceData,
+  VoiceFormData,
+  VoicePostData,
+} from 'types/course/assessment/question/voice-responses';
+
+import CourseAPI from 'api/course';
+import { JustRedirect } from 'api/types';
+
+export const fetchNewVoiceQuestion = async (): Promise<
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/question/voice-responses/operations.ts:11]: nit: `fetchNewVoiceResponse`

File: client/app/bundles/course/assessment/question/voice-responses/operations.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  VoiceFormData<'new'>
+> => {
+  const response = await CourseAPI.assessment.question.voice.fetchNewVoice();
+  return response.data;
+};
+
+export const fetchEditVoiceQuestion = async (
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/question/voice-responses/operations.ts:18]: nit: `fetchEditVoiceResponse`

File: client/app/bundles/course/assessment/question/voice-responses/NewVoicePage.tsx [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+import Preload from 'lib/components/wrappers/Preload';
+import useTranslation from 'lib/hooks/useTranslation';
+
+import translations from '../../translations';
+
+import { createVoiceQuestion, fetchNewVoiceQuestion } from './operations';
+import VoiceForm from './VoiceForm';
+
+const NEW_VOICE_TEMPLATE: VoiceData['question'] = {
+  title: '',
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/question/voice-responses/NewVoicePage.tsx:17]: You can abstract all these common fields out

File: client/app/bundles/course/assessment/question/components/CommonFieldsInQuestionForm.tsx [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+import FormRichTextField from 'lib/components/form/fields/RichTextField';
+import FormTextField from 'lib/components/form/fields/TextField';
+import useTranslation from 'lib/hooks/useTranslation';
+
+import translations from '../../translations';
+
+import SkillsAutocomplete from './SkillsAutocomplete';
+
+interface CommonFieldsProps {
+  disabled?: boolean;
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/question/components/CommonFieldsInQuestionForm.tsx:16]: Why are `undefined` added to all props?

File: spec/features/course/assessment/question/voice_response_management_spec.rb [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+      scenario 'I can edit a question' do
+        voice = create(:course_assessment_question_voice_response, assessment: assessment)
+        visit course_assessment_path(course, assessment)
+
+        edit_path = edit_course_assessment_question_voice_response_path(course, assessment, voice)
+        find_link(nil, href: edit_path).click
+
+        maximum_grade = 999.9
+        fill_in 'maximumGrade', with: maximum_grade
```
Comment (ekowidianto) [inline spec/features/course/assessment/question/voice_response_management_spec.rb:51]: Could you change other fields as well (i.e title, description, etc)

**Review by ekowidianto — CHANGES_REQUESTED** [review]
Thanks @bivanalhar
Please refer to my comments below.

File: app/controllers/course/assessment/question/voice_responses_controller.rb [hunk 13]
```diff
@@ -44,6 +42,16 @@ def destroy
 
   private
 
+  def update_skill_ids_if_params_present
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply app/controllers/course/assessment/question/voice_responses_controller.rb:45]: alright noted :) thanks for the feedback. I have refactored it accordingly

File: client/app/bundles/course/assessment/question/voice-responses/validation.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import { array, number, object, string } from 'yup';
+
+import translations from '../../translations';
+
+const questionSchema = object({
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/assessment/question/voice-responses/validation.ts:5]: alright noted :) thanks for the feedback. I have refactored it accordingly

File: app/views/course/assessment/question/voice_responses/edit.json.jbuilder (hunk 2 above)
Comment (bivanalhar) [reply app/views/course/assessment/question/voice_responses/edit.json.jbuilder:6]: after I understand what it actually entails, I concluded it's not needed. I have removed it in the commit that will follow

File: app/views/course/assessment/question/voice_responses/new.json.jbuilder (hunk 1 above)
Comment (bivanalhar) [reply app/views/course/assessment/question/voice_responses/new.json.jbuilder:3]: after I understand what it actually entails, I concluded it's not needed. I have removed it in the commit that will follow

File: client/app/api/course/Assessment/Question/index.ts (hunk 4 above)
Comment (bivanalhar) [reply client/app/api/course/Assessment/Question/index.ts:8]: done

File: client/app/types/course/assessment/question/voice-responses.ts (hunk 7 above)
Comment (bivanalhar) [reply client/app/types/course/assessment/question/voice-responses.ts:14]: I'm really sorry. At first I thought it's the abstraction to allow randomization of skills (I know it sounds a bit stupid).. apparently after I re-read it again, it's specific for MCQ/MRQ and thus I have deleted it

File: client/app/bundles/course/assessment/question/voice-responses/operations.ts (hunk 8 above)
Comment (bivanalhar) [reply client/app/bundles/course/assessment/question/voice-responses/operations.ts:11]: done!!

File: client/app/bundles/course/assessment/question/voice-responses/operations.ts (hunk 9 above)
Comment (bivanalhar) [reply client/app/bundles/course/assessment/question/voice-responses/operations.ts:18]: done!!

File: client/app/bundles/course/assessment/question/voice-responses/NewVoicePage.tsx (hunk 10 above)
Comment (bivanalhar) [reply client/app/bundles/course/assessment/question/voice-responses/NewVoicePage.tsx:17]: done!!

File: spec/features/course/assessment/question/voice_response_management_spec.rb (hunk 12 above)
Comment (bivanalhar) [reply spec/features/course/assessment/question/voice_response_management_spec.rb:51]: I have tried, but it can only be tested after PR.. so what I will do now shall be to commit then push it here.. hopefully all goes well

File: client/app/bundles/course/assessment/question/multiple-responses/NewMcqMrqPage.tsx [hunk 15]
```diff
@@ -27,15 +28,21 @@ const newMcqMrqAdapters: Record<McqMrqFormData['mcqMrqType'], Adapter> = {
...
-  description: '',
-  staffOnlyComments: '',
-  maximumGrade: '',
-  skillIds: [],
+  ...newCommonTemplate,
   skipGrading: false,
   randomizeOptions: false,
 };
 
+// const NEW_MCQ_MRQ_TEMPLATE: McqMrqData['question'] = {
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/question/multiple-responses/NewMcqMrqPage.tsx:36]: Remove comment

**Review by ekowidianto — APPROVED** [review]

File: spec/features/course/assessment/question/multiple_response_management_spec.rb [hunk 16]
```diff
@@ -139,10 +139,10 @@
 
         visit edit_path
         options.each do |option|
-          click_button 'Add a new response'
+          click_button 'Add a new choice'
```
Comment (ekowidianto) [inline spec/features/course/assessment/question/multiple_response_management_spec.rb:142]: Why did you change this?

File: app/controllers/course/assessment/question/controller.rb [hunk 17]
```diff
@@ -25,6 +25,12 @@ def load_question_assessment_for(question)
     @assessment.question_assessments.find_by!(question: question.acting_as)
   end
 
+  def update_skill_ids_if_params_present
+    question_assessment_params = voice_response_question_params[:question_assessment]
```
Comment (ekowidianto) [inline app/controllers/course/assessment/question/controller.rb:29]: `voice_response_question_params` should have been updated too.. I will change it for you

File: spec/features/course/assessment/question/multiple_response_management_spec.rb (hunk 16 above)
Comment (bivanalhar) [reply spec/features/course/assessment/question/multiple_response_management_spec.rb]: sorry I looked into the CircleCI and they indicated this testcase was not passed. So I tried to change it, but apparently also still not pass. Now I have reverted it back

File: app/controllers/course/assessment/question/controller.rb (hunk 17 above)
Comment (bivanalhar) [reply app/controllers/course/assessment/question/controller.rb]: ah sorry, previously I accidentally force push the update from my local to here, so maybe you need to check if the current version is reverted back..

Though, I have refactored this one. So you might need to check on this again

Comment (ekowidianto) [reply app/controllers/course/assessment/question/controller.rb]: You have just overriden my commit...
I would suggest using force-with-lease instead of just force next time

Comment (bivanalhar) [reply app/controllers/course/assessment/question/controller.rb]: I'm sorry :( will not happen again next time. and sure noted on that.

Comment (ekowidianto) [reply app/controllers/course/assessment/question/controller.rb]: Dont worry :)

File: spec/features/course/assessment/question/multiple_response_management_spec.rb (hunk 16 above)
Comment (ekowidianto) [reply spec/features/course/assessment/question/multiple_response_management_spec.rb]: You can always try it locally after you see if theres any failed test cases in circleci
